src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs

```diff
@@ -109,18 +110,43 @@ private static CmsSigner CreateCmsSigner(SignPackageRequest request)
                 signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, request.Certificate);
             }
 
-            var chain = SigningUtility.GetCertificateChain(request.Certificate, request.AdditionalCertificates);
-
-            foreach (var certificate in chain)
-            {
-                signer.Certificates.Add(certificate);
-            }
-
-            var attributes = SigningUtility.GetSignAttributes(request, chain);
-
-            foreach (var attribute in attributes)
+            using (var chain = new X509Chain())
             {
-                signer.SignedAttributes.Add(attribute);
+                SigningUtility.SetCertBuildChainPolicy(
+                    chain.ChainPolicy,
+                    request.AdditionalCertificates,
+                    DateTime.Now,
+                    NuGetVerificationCertificateType.Signature,
+                    NuGetChainBuildingRequestType.Signing);
+
+                if (SigningUtility.BuildCertificateChain(chain, request.Certificate, out var chainStatusList))
+                {
+                    var chainList = SigningUtility.CertificateChainToList(chain);
+
+                    foreach (var certificate in chainList)
+                    {
+                        signer.Certificates.Add(certificate);
+                    }
+
+                    var attributes = SigningUtility.GetSignAttributes(request, chainList);
+
+                    foreach (var attribute in attributes)
+                    {
+                        signer.SignedAttributes.Add(attribute);
+                    }
+                }
+                else
+                {
+                    var messages = chainStatusList
+                        .ToList()
+                        .Select(x => x.StatusInformation?.Trim())
+                        .Where(x => !string.IsNullOrEmpty(x))
+                        .Distinct()
+                        .OrderBy(x => x)
+                        .ToList();
```

Comment:
nit: ToList is useful to avoid enumerating and running everything multiple times. Here it looks like this would only get enumerated by string.Join, so you really don't need to create a list. These linq statements are setting up a query, and string.Join will execute it only once which is the same as ToList executing it once. The difference is that you don't need to allocate space for the list itself if you let string.Join enumerate it.